### PR TITLE
DWPF-778 Show bad results message

### PR DESCRIPTION
**Pull request by nicopicchio**

### CODE REVIEW

File: src/search/search.py [hunk 1]
```diff
@@ -7,20 +7,16 @@
 
 
 class SearchVector:
-    def __init__(self, request, annotate_score=False):
+    def __init__(self, request):
         self.request = request
-        self.annotate_score = annotate_score
+        self.annotate_score = True
```
Comment (CamLamb) [inline src/search/search.py:12]: `self.annotate_score` is no longer used anywhere

File: src/search/utils.py [hunk 2]
```diff
@@ -136,3 +142,86 @@ def split_query(query: str) -> list[str]:
...
+    boost_values = set()
+    subqueries = get_all_subqueries(query)
+
+    for subquery in subqueries[category]:
+        boost_values.add(round(subquery["boost"], 2))
+
+    return statistics.median(boost_values) * bad_score_threshold_multiplier
+
+
+def query_has_bad_results(query, category, pinned_results, search_results):
```
Comment (CamLamb) [inline src/search/utils.py:218]: Can you add a comment to explain the purpose of this method?

File: src/search/utils.py [hunk 3]
```diff
@@ -136,3 +142,86 @@ def split_query(query: str) -> list[str]:
...
+            query, Person, mapping
+        )
+        get_query_info(subqueries["people"], field, mapping, analyzer_field_suffices)
+    for mapping in Team.IndexManager.get_mapping():
+        field = Team.IndexManager._get_search_query_from_mapping(query, Team, mapping)
+        get_query_info(subqueries["teams"], field, mapping, analyzer_field_suffices)
+    return subqueries
+
+
+def get_bad_score_threshold(query, category):
```
Comment (CamLamb) [inline src/search/utils.py:207]: Can you add a comment to explain the purpose of this method?

File: src/search/templates/search/partials/search_results_category.html [hunk 4]
```diff
@@ -1,4 +1,5 @@
 <div class="search-results__category">
+    {% if show_bad_results_message %}<h1>These results might not be what you are looking for</h1>{% endif %}
```
Comment (marcelkornblum) [inline src/search/templates/search/partials/search_results_category.html:2]: I think we need a few different versions of this copy depending on other criteria - are there lots of results or only a few? which type of search are we on?

File: src/search/utils.py (hunk 2 above)
Comment (marcelkornblum) [inline src/search/utils.py:218]: could we rename this to be more on-the-nose please? maybe `only_has_bad_results` or `has_no_good_results` or something?

File: src/search/utils.py [hunk 5]
```diff
@@ -1,6 +1,12 @@
 import re
 import unicodedata
+import statistics
```
Comment (marcelkornblum) [inline src/search/utils.py:3]: i tihnk this is overkill for what we want to do

File: src/search/utils.py [hunk 6]
```diff
@@ -136,3 +142,86 @@ def split_query(query: str) -> list[str]:
...
+        )
+        get_query_info(subqueries["people"], field, mapping, analyzer_field_suffices)
+    for mapping in Team.IndexManager.get_mapping():
+        field = Team.IndexManager._get_search_query_from_mapping(query, Team, mapping)
+        get_query_info(subqueries["teams"], field, mapping, analyzer_field_suffices)
+    return subqueries
+
+
+def get_bad_score_threshold(query, category):
+    bad_score_threshold_multiplier = 1
```
Comment (marcelkornblum) [inline src/search/utils.py:208]: this should be extracted to an easier-to-change setting. I tihnk an ENV var in settings is probably best.

File: src/search/utils.py [hunk 7]
```diff
@@ -136,3 +142,86 @@ def split_query(query: str) -> list[str]:
...
+    boost_values = set()
+    subqueries = get_all_subqueries(query)
+
+    for subquery in subqueries[category]:
+        boost_values.add(round(subquery["boost"], 2))
+
+    return statistics.median(boost_values) * bad_score_threshold_multiplier
```
Comment (marcelkornblum) [inline src/search/utils.py:215]: i am not convinced median is right - more like the 70th percentile maybe, just as the curve gets steeper? either way I don't think we need the stats package here; all we're doing is choosing the nth value where N / total = {our percentile}. so if we want median we total the number of boosts, halve the total to get N and select the boost at position N in the list.

File: src/search/utils.py [hunk 8]
```diff
@@ -136,3 +142,86 @@ def split_query(query: str) -> list[str]:
         parts.append(group)
 
     return parts
+
+
+def get_query_info(fields, field, mapping, suffix_map):
```
Comment (marcelkornblum) [inline src/search/utils.py:147]: these copied-acrss functions would ideally get tidied up, they're such a mess. not going to mind if you don't though :)

**Review by marcelkornblum — CHANGES_REQUESTED** [review]
Good as far as it goes, but there are a few bits missing from what the ticket needs (see specific comments)

File: src/search/templates/search/partials/search_results_category.html (hunk 4 above)
Comment (nicopicchio) [reply src/search/templates/search/partials/search_results_category.html:2]: we do but I am waiting for Bobby to come up with a final copy. initially it was `hello world`, I've only changed it to give a more meaningful message before opening this PR

Comment (marcelkornblum) [reply src/search/templates/search/partials/search_results_category.html:2]: sure, but it also needs to be more than one message depending on other factors - so those factors ought to make an appearance here

File: src/search/utils.py (hunk 5 above)
Comment (nicopicchio) [reply src/search/utils.py:3]: ok, what do you suggest instead?

Comment (marcelkornblum) [reply src/search/utils.py:3]: commented below

File: src/search/utils.py (hunk 2 above)
Comment (nicopicchio) [reply src/search/utils.py:218]: renamed to `has_only_bad_results`

File: src/search/utils.py (hunk 6 above)
Comment (nicopicchio) [reply src/search/utils.py:208]: done

File: src/search/templatetags/search.py [hunk 9]
```diff
@@ -33,19 +34,26 @@
...
 )
 def search_category(
-    context, *, category, tab_name=None, limit=None, show_heading=False
+    context,
+    *,
+    category,
+    tab_name=None,
+    limit=None,
+    show_heading=False,
+    show_bad_threshold_message=True,
```
Comment (marcelkornblum) [inline src/search/templatetags/search.py:43]: could we rationalise the variables `show_bad_threshold_message` and `show_bad_results_message` across the whole codebase please? I can't see a reason to have them being different and it'd be easier to understand the code if they were named the same (assuming they are in fact doing the same thing)

**Review by marcelkornblum — COMMENTED** [review]
just one last thing please - otherwise looking fantastic

File: src/search/templatetags/search.py (hunk 9 above)
Comment (nicopicchio) [reply src/search/templatetags/search.py]: renamed `show_bad_results_message`

**Pull request merged**